cxx/distributions/bigram.hh

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#pragma once
+#include "base.hh"
+#include "dirichlet_categorical.hh"
+
+class Bigram : public Distribution<std::string> {
+private:
+    void assert_valid_char(const char c) const {
+        assert(c >= ' ' && c <= '~');
+    }
+    size_t char_to_index(const char c) const {
+        assert_valid_char(c);
+        return c - ' ';
+    }
+    char index_to_char(const size_t i) const {
+        const char c = i + ' ';
+        assert_valid_char(c);
+        return c;
+    }
+    std::vector<size_t> string_to_indices(const std::string& str) const {
+        // Convert the string to a vector of indices between 0 and `num_chars`,
+        // with a start/stop symbol at the beginning/end.
+        std::vector<size_t> inds = {num_chars};
+        for (const char& c : str) {
+            inds.push_back(char_to_index(c));
+        }
+        inds.push_back(num_chars);
+        return inds;
+    }
+public:
+    double alpha = 1;  // hyperparameter for all transition distributions.
+    size_t num_chars = '~' - ' ' + 1;  // printable ASCII without DEL.
+    std::vector<DirichletCategorical> transition_dists;
+    PRNG *prng;
+
+    Bigram(PRNG *prng) {
+        this->prng = prng;
+        const size_t total_chars = num_chars + 1;  // Include a start/stop symbol.
+
+        // The distribution at index `i` represents `p(x_{i+1} | x_i)`.
+        transition_dists.reserve(total_chars);
+        for (size_t i = 0; i != total_chars; ++i) {
+            transition_dists.emplace_back(prng, total_chars);
+        }
+    }
+    void incorporate(const std::string& x) {
+        const std::vector<size_t> indices = string_to_indices(x);
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            transition_dists[indices[i]].incorporate(indices[i + 1]);
+        }
+    }
+    void unincorporate(const std::string& s) {
+        const std::vector<size_t> indices = string_to_indices(s);
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            transition_dists[indices[i]].unincorporate(indices[i + 1]);
+        }
+    }
+    double logp(const std::string& s) const {
+        const std::vector<size_t> indices = string_to_indices(s);
+        double total_logp = 0.0;
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            total_logp += transition_dists[indices[i]].logp(indices[i + 1]);
```

Comment:
So, the one thing that this strategy won't account for is that in a string that repeats the same transition multiple times, each time it happens it makes the next one more likely.

I think a strategy that could work would be to, after line 64 (but still in the body of the for loop), incorporate the observed transition in the underlying transition_dist. This will ensure that on future iterations of the loop, the logp's are being properly computed based on previous iterations. Then, at the end of the for loop, call unincorporate on `s` (which will undo all the incorporating).

(A wrinkle is that you may not be allowed to call a non-`const` method like incorporate/unincorporate from a `const` method like `logp`? Not sure how best to handle this.)

Reply:
Thanks, I hadn't thought of that. I marked `transition_dists` as `mutable` and copy/pasted the (short) body of `unincorporate` at the end of `logp`. 

I also thought about just making `logp` non-`const`, since it appears to never be called on `const` distributions. (Also, it isn't clear to me why `logp` is `const` and `sample` isn't.)

Reply:
Sounds good! And yes, I also wondered that. I thought maybe it had to do with PRNG state being mutated... but not sure.